plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/CallSiteInline.hs

```diff
@@ -0,0 +1,123 @@
+{-# LANGUAGE ConstraintKinds  #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTs            #-}
+{-# LANGUAGE TypeFamilies     #-}
+
+{-|
+Call site inlining machinery. For now there's only one part: inlining of fully applied functions.
+See ADR TODO for motivation. We inline fully applied functions if the cost and size are acceptable.
+See note [Inlining of fully applied functions].
+-}
+
+module PlutusIR.Transform.Inline.CallSiteInline where
+
+{- Note [Inlining of fully applied functions]
+
+We inline the call site when a function is fully applied AND the cost and size are acceptable.
+
+Consider `let v :: ty = rhs in body`, in which `body` calls `v`.
+
+We consider cases when `v` is a function/lambda abstraction(s). I.e.:
+
+let v :: type = \x1.\x2...\xn.VBody in body
+
+In the `body`, where `v` is *called*,
+if it was given `n` arguments, then it is fully applied in the `body`.
+We inline the *call site* of `v` in this case, i.e.,
+we replace `v x_1 ... x_n` in the `body` with `rhs`. E.g.
+
+let f :: TyLam (TODO)= \x.\y -> x*y
+    y :: TyLam = f q
+in f a b
+
+becomes `let y = f q in a*b`. What about
```

Comment:
1. You still need the `let f :: ...` otherwise `f` is undefined here
2. You may want to replace `a * b` with something smaller, say `a`. Because `a * b` and `f a b` have the same size, so it is unclear why we want to turn `f a b` into `a * b`.